dev/src/transaction.ts

```diff
@@ -609,6 +671,8 @@ function isRetryableTransactionError(error: GoogleError): boolean {
         // IDs that have expired. While INVALID_ARGUMENT is generally not
         // retryable, we retry this specific case.
         return !!error.message.match(/transaction has expired/);
+      case StatusCode.FAILED_PRECONDITION:
```

Comment:
question: why do we retry FAILED_PRECONDITION for optimistic locking?

Reply:
There should already be a comment (in the last diff).